watertap/tools/analysis_tools/loop_tool/loop_tool.py

```diff
@@ -0,0 +1,677 @@
+                                    self.expected_num_samples
+                                ),
+                                "force_rerun": copy.deepcopy(self.force_rerun),
+                                "min_num_samples": copy.deepcopy(self.min_num_samples),
+                                "diff_params": copy.deepcopy(self.diff_params),
+                                "diff_samples": copy.deepcopy(self.diff_samples),
+                                "original_options_dict": copy.deepcopy(self.options),
+                            }
+                        }
+                    else:
+                        print("loop empty did not add", loop_value)
+        return sweep_directory, cur_dir
+
+    def update_dir_path(self, cur_dir, key, value):
+        """creates directory pathing for h5 file strucutre"""
+        if value == None:
+            cur_dir = cur_dir + "/" + str(key)
+        else:
+            if cur_dir.find(key) > 0:
+                cur_dir = cur_dir[: (cur_dir.find(key) - 1)]
+            cur_dir = cur_dir + "/" + str(key)
+            cur_dir = cur_dir + "/" + str(value)
+        return cur_dir
+
+    def update_sim_options(self, loop_type, loop_key, loop_value, loop):
+        """used to update simulation options for each specific loop"""
+
+        if loop_type == "build_loop":
+            self.build_defaults.update(self.get_loop_params(loop_key, loop_value, loop))
+        elif loop_type == "init_loop":
+            self.init_defaults.update(self.get_loop_params(loop_key, loop_value, loop))
+        elif loop_type == "optimize_loop":
+            self.optimize_defaults.update(
+                self.get_loop_params(loop_key, loop_value, loop)
+            )
+        elif loop_type == "sweep_param_loop":
+            self.sweep_params = {}
+            (
+                self.sweep_params,
+                self.num_samples,
+                self.expected_num_samples,
+                self.min_num_samples,
+                self.force_rerun,
+            ) = self.get_sweep_params(loop_key, loop_value)
+
+        elif loop_type == "diff_param_loop":
+            self.sweep_params = {}
+            self.diff_params = {}
+            if loop_key != "sweep_reference_params":
+                (
+                    self.sweep_params,
+                    self.num_samples,
+                    self.diff_params,
+                    self.diff_samples,
+                    self.expected_num_samples,
+                    self.min_num_samples,
+                    self.force_rerun,
+                ) = self.get_diff_params(loop_key, loop_value)
+
+    def get_loop_params(self, key, loop_value, loop):
+        if "case" in key:
+            return loop[key][loop_value]
+        elif ":" in str(loop_value):
+            lp = loop_value.split(":")
+            return {key: {eval(lp[0]): float(lp[1])}}
+        else:
+            return {key: loop_value}
+
+    def setup_multi_processing(self):
```

Comment:
We should use the parallel manager to setup parallel computing. It has logic for how to handle concurrent futures or MPI depending on the user specification and conda environment.

Reply:
This is for the loopTool, as it needs to know if its running with MPI or not for when handling file saving (e.g. getting rank 0), other wise it setups number of subprocess and passes it to parameter sweep tool, which sets up its own parallel management.